libs/MeshKernel/src/CurvilinearGrid/CurvilinearGridFromSplines.cpp

```diff
@@ -592,11 +645,335 @@ namespace meshkernel
+            Point x4 = splinePoints[i + 1];
+
+            if (!x3.IsValid() || !x4.IsValid())
+            {
+                continue;
+            }
+
+            const auto [segmentsDoCross,
+                        intersection,
+                        crossProduct,
+                        firstRatio,
+                        secondRatio] = AreSegmentsCrossing(x1, x2, x3, x4, false, m_splines->m_projection);
+
+            if (segmentsDoCross)
+            {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    void CurvilinearGridFromSplines::SolveQuartic(const std::array<double, 5>& coefficients,
+                                                  std::array<double, 4>& roots) const
+    {
+        int degree = 4;
+        std::array<double, 5> coeffs(coefficients);
+        std::array<double, 4> realRoots{0.0, 0.0, 0.0, 0.0};
+        std::array<double, 4> imagRoots{0.0, 0.0, 0.0, 0.0};
+        roots.fill(constants::missing::doubleValue);
+
+        for (UInt i = 4; i >= 1; --i)
+        {
+
+            if (std::abs(coeffs[4 - i]) > tolerance)
+            {
+                degree = i;
+                break;
+            }
+        }
+
+        // shuffle coefficients up if necessary
+        if (degree < 4)
+        {
+            for (int j = 0; j <= degree; ++j)
+            {
+                coeffs[j] = coeffs[4 - degree + j];
+            }
+
+            for (int j = degree + 1; j < 5; ++j)
+            {
+                coeffs[j] = 0.0;
+            }
+        }
+
+        RootFinder(coeffs, degree, realRoots, imagRoots);
+
+        if (degree <= 0)
+        {
+            return;
+        }
+
+        for (UInt i = 0; i < static_cast<UInt>(degree); ++i)
+        {
+            if (std::abs(imagRoots[i]) < 1.0e-4)
+            {
+                roots[i] = realRoots[i];
+            }
+        }
+    }
+
+    double CurvilinearGridFromSplines::ComputeNodeSegmentCrossingTime(const Point& x1, const Point& x3, const Point& x4,
+                                                                      const Point& v1, const Point& v3, const Point& v4) const
+    {
+        if ((x1.y - x3.y) * (x4.x - x3.x) - (x1.x - x3.x) * (x4.y - x3.y) < 0.0)
+        {
+            return 1.0e99;
+        }
+
+        // Compute differences
+        Point x13 = x3 - x1;
+        Point x34 = x4 - x3;
+        Point v13 = v3 - v1;
+        Point v34 = v4 - v3;
+
+        double a = cross(v13, v34);
+        double b = cross(x13, v34) - cross(x34, v13);
+        double c = cross(x13, x34);
+
+        std::array<double, 5> coeffs{0.0, 0.0, a, b, c};
+        std::array<double, 4> roots{0.0, 0.0, 0.0, 0.0};
+        std::array<double, 4> beta;
+        beta.fill(constants::missing::doubleValue);
+        roots.fill(constants::missing::doubleValue);
+
+        SolveQuartic(coeffs, roots);
+
+        for (UInt i = 0; i < 4; ++i)
+        {
+
+            if (roots[i] == constants::missing::doubleValue)
```

Comment:
I really do not think that we should check if something is approximately the missing value, it must be exactly the missing value. This value will not be calculated from something else, it will be assigned and not changed.